### ISSUE TITLE
Feature: Allow for unformatted output of failures

### ISSUE THREAD

**Issue by rundis** [elm-community/elm-test#74]
## TLDR;

Allow facility to retrieve detailed information about a test failure
## Context

I'm currently working on a PR for allowing [node-test-runner](https://github.com/rtfeldman/node-test-runner) to output JSON as an optional format for test reporting. Reason being that I'm trying to do some small steps to make elm-test more tooling(/editor) friendly.
## Current design limitations

It's great to have consistent output from elm-test across runners. However for tooling integration it's not necessarily so great that interested detailed information about failures get's baked into a formatted string. You could argue that tools should always show errors the way elm-test formats them. However I think there is a pretty good case for allowing some flexibility and room for innovative ways of showing test results. (Maybe hightlight diffs,  use levenshtein for smarter text diffs or colouring to highlight mismatches for collections etc) 
## Suggestion
- Don't format stuff when reporting the failures
- Format upon retrieving failure information
- Add new function(s) to retrieve raw failure information

I can totally see some challenges with this, especially with the collections reporting. But still feel that delaying (optional) output formatting to the point of retrieval would be a sounder design in the long run though. It will also fall nicely in line with how labels can be optionally formatted with `formatLabels`

**Comment by mgold** [comment]
It seems reasonable for the node runner to have a `--json` flag or something. I haven't really thought through how this would affect the semi-private functions that run the test; probably some intermediate representation that can become either JSON or a string?

**Comment by rtfeldman** [comment]
Closed by https://github.com/rtfeldman/node-test-runner/pull/81

**Comment by rundis** [comment]
I assume this was closed because I wasn't clear in communicating what I was suggesting here ? 
Maybe I should open a new issue with a clearer description ?

**tldr; Allow greater flexibility in how one may report failures**

What I was trying to get at was really boils down to the limitations the design the Expectation type (and it's usage):
https://github.com/elm-community/elm-test/blob/master/src/Test/Expectation.elm#L6

With the current design failures are formatted to strings and wrapped in the `Fail String String` at the point in time when the expectations are verified/run. By doing so there is no room for error reporters (like the new json reporter) to provide alternative formatting or output of the failure details). It's all baked into a string. 
My suggestion is basically to either:
1) Change `Expectation.Fail` to contain the raw values of expected value and actual value
2) Extend `Expectation.Fail` to contain the raw values of expected and actual

Then in addition (depending on alternative chosen above)
1) Defer formatting of output to the point in time the user calls [Expectation.getFailure](https://github.com/elm-community/elm-test/blob/master/src/Expect.elm#L512)
2) Provide a new function `Expectation.getFailureExtended` or `Expectation.getFailureRaw` or something like that, which (also) provides the raw expected and actual values.
Of course to use these results there is a pretty interesting challenge in terms of also representing the types of expected and actual.

Assuming the above addition is even possible (or feasible without complecting elm-test in a way that wouldn't justify any conceivable benefits of the change, reporters would now have much higher degree in flexibility in how they would like to report failures.

Maybe with aomething to the effect as outlined above the json reporter in node test runner could 
jsonify actual and expected so that tooling could do much more sophisticated presentation of failure diffs. Maybe a html runner could present results in much cooler/helpful ways etc.

Idk if you think this is worth exploring further or not ?

**Comment by mgold** [comment]
@rundis I've felt that the representation of failures was too simplistic for awhile now. Richard and I worked on this together at the SF meetup and #85 is our first pass. It pretty much follows your suggestion: enrich `Expectation.Fail`, and don't do vertical bar formatting until `Expect.getFailure`. And yes, this opens the door to adding another function with a more enriched data output, though we'd want to do it in a way that allows us to keep extending `Fail`. For example, adding `Expect.between` or `Expect.custom` would need another addition to the internal union type `Reason`.
